VERSIONING.md

```diff
@@ -0,0 +1,128 @@
+# Versioning
+
+This document covers Vector's versioning and what it means as a user of Vector.
+
+**Please note, Vector is currently in its pre-1.0 phase and quickly approaching
+1.0. Minor version increments can introduce breaking changes during this phase.
+Please see the [what to expect](#what-to-expect) section for more info.**
+
+<!-- MarkdownTOC autolink="true" style="ordered" indent="   " -->
+
+1. [Convention](#convention)
+1. [Public API](#public-api)
+   1. [Areas that *are* covered](#areas-that-are-covered)
+      1. [Intended for *public* consumption](#intended-for-public-consumption)
+      1. [Intended for *private* consumption](#intended-for-private-consumption)
+   1. [Areas that are *NOT* covered](#areas-that-are-not-covered)
+1. [What to expect](#what-to-expect)
+   1. [Patch and minor versions](#patch-and-minor-versions)
+   1. [Major versions \(breaking changes\)](#major-versions-breaking-changes)
+   1. [Pre 1.0 version](#pre-10-version)
+
+<!-- /MarkdownTOC -->
+
+## Convention
+
+Vector adheres to the [Semantic Versioning 2.0] convention. In summary:
+
+* Versions follow the `MAJOR.MINOR.PATCH` format (i.e., `2.5.1`)
+* `PATCH` increments only when backward compatible bug fixes are introduced.
+* `MINOR` increments only when new, backward compatible functionality is introduced.
+* `MAJOR` increments if any backwards incompatible changes are introduced.
+* Pre `1.0` (major version `0`) is for initial development and `MINOR` version bumps can introduce breaking changes.
+
+## Public API
+
+Semantic Versioning hinges on Vector's defintion of "public API". By nature of
+Vector - a tool that collects, processes, and routes data from disparate systems
+- it has a very large public surface area. It's not immediately obvious which
+parts are covered under our versioning contract and how they're covered. This
+section aims to remove all ambiguity in this area.
+
+### Areas that *are* covered
+
+The following Vector areas are covered in Vector's definition of public API.
+
+#### Intended for *public* consumption
```

Comment:
Fixed in https://github.com/timberio/vector/pull/8795/commits/22191284121eebc53d0107b82814ea0635548253

Reply:
I'm still not seeing internal metrics mentioned here or the sink encodings (I'm thinking like `encoding.codec = "json"` always encodes events in the same way).

Reply:
Are encoding covered? We don't guarantee anything about broad JSON encoding except that it encodes to JSON. I addressed the `vector` sink separately.

Reply:
> We don't guarantee anything about broad JSON encoding except that it encodes to JSON

We have been guaranteeing this so far at least. If we change it, it could break whatever upstream components Vector is sending to. For example: a remote HTTP sink.

Reply:
Granted, for logs, it just means that all fields are encoded as top-level JSON fields, but this should still be a guarantee that we don't switch to, for example, `{ "attrs": { all fields } }`, without a major (or minor, pre-1.0) version bump.

Reply:
Makes sense. Fixed in https://github.com/timberio/vector/pull/8795/commits/80959eab4aea3d6215324e210c3ce19dd7b28e37.